File: server/fcm_server.py
from fastapi import FastAPI, APIRouter, HTTPException
from pydantic import BaseModel
import firebase_admin
from firebase_admin import credentials, firestore, messaging
import os
import asyncio
import uvicorn
from fastapi.middleware.cors import CORSMiddleware # 引入 CORSMiddleware

# 確保 serviceAccountKey.json 存在
# 這裡的路徑需要根據 fcm_server.py 相對於 serviceAccountKey.json 的位置來調整
# 假設 serviceAccountKey.json 也在 server/ 目錄下
SERVICE_ACCOUNT_KEY_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'serviceAccountKey.json')

if not os.path.exists(SERVICE_ACCOUNT_KEY_PATH):
    print(f"錯誤：找不到服務帳戶金鑰檔案：{SERVICE_ACCOUNT_KEY_PATH}")
    raise FileNotFoundError("serviceAccountKey.json not found")

# 初始化 Firebase Admin SDK
if not firebase_admin._apps:
    cred = credentials.Certificate(SERVICE_ACCOUNT_KEY_PATH)
    firebase_admin.initialize_app(cred)
    print("Firebase Admin SDK initialized successfully for FCM server.")

db = firestore.client()

app = FastAPI(
    title="FCM Management API",
    description="API for managing FCM registrations and topic subscriptions.",
    version="1.0.0",
)

# 添加 CORS 中間件
app.add_middleware(
    CORSMiddleware,
    allow_origins=["http://localhost:9000"], # 允許前端的來源
    allow_credentials=True,
    allow_methods=["*"], # 允許所有 HTTP 方法
    allow_headers=["*"], # 允許所有請求頭
)

class FcmRegistration(BaseModel):
    uid: str
    fcmToken: str
    townshipCode: str | None = None

async def subscribe_to_weather_topic(fcm_token: str, township_code: str):
    topic = f"weather_{township_code}"
    try:
        response = messaging.subscribe_to_topic([fcm_token], topic)
        print(f"Successfully subscribed token {fcm_token} to topic {topic}: {response}")
        return True
    except Exception as e:
        print(f"Error subscribing token {fcm_token} to topic {topic}: {e}")
        return False

async def unsubscribe_from_weather_topic(fcm_token: str, topic: str):
    try:
        response = messaging.unsubscribe_to_topic([fcm_token], topic)
        print(f"Successfully unsubscribed token {fcm_token} from topic {topic}: {response}")
        return True
    except Exception as e:
        print(f"Error unsubscribing token {fcm_token} from topic {topic}: {e}")
        return False

@app.post("/api/fcm/register")
async def register_fcm(registration: FcmRegistration):
    try:
        user_ref = db.collection("fcmTokens").document(registration.uid)
        user_doc = await asyncio.to_thread(user_ref.get) # Firestore get is synchronous

        old_fcm_token = None
        old_township_code = None
        old_topic = None

        if user_doc.exists:
            data = user_doc.to_dict()
            old_fcm_token = data.get("fcmToken")
            old_township_code = data.get("townshipCode")
            if old_township_code:
                old_topic = f"weather_{old_township_code}"

        # 1. 處理 Token 變更或 TownshipCode 變更導致的訂閱更新
        if old_fcm_token != registration.fcmToken: # Token 變更，需要重新訂閱
            if old_fcm_token and old_topic: # 如果有舊 Token 和舊 Topic，先取消訂閱
                await unsubscribe_from_weather_topic(old_fcm_token, old_topic)
            if registration.townshipCode: # 用新 Token 訂閱新 Topic
                await subscribe_to_weather_topic(registration.fcmToken, registration.townshipCode)
        elif registration.townshipCode and old_township_code != registration.townshipCode: # Token 沒變，但 TownshipCode 變了
            if old_topic: # 取消訂閱舊 Topic
                await unsubscribe_from_weather_topic(registration.fcmToken, old_topic)
            await subscribe_to_weather_topic(registration.fcmToken, registration.townshipCode) # 訂閱新 Topic
        elif registration.townshipCode and not old_township_code: # 第一次設定 TownshipCode
            await subscribe_to_weather_topic(registration.fcmToken, registration.townshipCode)
        
        # 2. 更新 Firestore 儲存
        await asyncio.to_thread(user_ref.set, {
            "uid": registration.uid,
            "fcmToken": registration.fcmToken,
            "townshipCode": registration.townshipCode,
            "lastUpdated": firestore.SERVER_TIMESTAMP
        })

        return {"message": "FCM registration and topic subscription updated successfully"}
    except Exception as e:
        print(f"FCM 註冊失敗: {e}")
        raise HTTPException(status_code=500, detail=f"FCM registration failed: {e}")

if __name__ == "__main__":
    try:
        uvicorn.run(app, host="0.0.0.0", port=7800)
    except Exception as e:
        print(f"FCM 伺服器啟動失敗: {e}")

File: server/fcm-server.js
// --- 引入必要的套件 ---
const express = require('express');
const admin = require('firebase-admin');
const cors = require('cors');

// --- 初始化 Firebase Admin SDK ---
// 請確認 serviceAccountKey.json 檔案與此檔案在同一個資料夾
const serviceAccount = require('./serviceAccountKey.json');

admin.initializeApp({
  credential: admin.credential.cert(serviceAccount)
});

const db = admin.firestore();
const messaging = admin.messaging();
const app = express();

// --- 中介軟體 (Middleware) ---
// 使用 CORS 中介軟體，並設定允許的來源
app.use(cors({
  origin: 'https://taiwan-weather-alert.pages.dev'
}));
// 使用 Express 內建的中介軟體來解析 JSON 格式的請求 Body
app.use(express.json());


// --- API 路由 (Endpoint) ---

/**
 * 根路由，用於健康檢查
 */
app.get('/', (req, res) => {
  res.status(200).send({ status: 'ok', message: 'FCM server (Node.js) is running.' });
});

/**
 * 處理 FCM Token 註冊與 Topic 訂閱的路由
 */
app.post('/api/fcm/register', async (req, res) => {
  console.log(`收到來自 ${req.ip} 的註冊請求`);
  
  // 從請求的 Body 中獲取資料
  const { uid, fcmToken, townshipCode } = req.body;

  // 驗證收到的資料
  if (!uid || !fcmToken) {
    console.error('請求資料不完整: 缺少 uid 或 fcmToken');
    return res.status(422).send({ error: 'uid and fcmToken are required.' });
  }
  
  console.log(`收到的資料: uid=${uid}, fcmToken=${fcmToken.substring(0, 10)}..., townshipCode=${townshipCode}`);

  try {
    const userRef = db.collection('fcmTokens').doc(uid);
    const userDoc = await userRef.get();

    const oldData = userDoc.exists ? userDoc.data() : {};
    const oldFcmToken = oldData.fcmToken;
    const oldTownshipCode = oldData.townshipCode;

    // 判斷是否需要更新訂閱
    if (oldFcmToken !== fcmToken || oldTownshipCode !== townshipCode) {
      console.log('偵測到 Token 或鄉鎮代碼變更，開始處理訂閱更新...');

      // 1. 如果有舊的訂閱，先取消
      if (oldFcmToken && oldTownshipCode) {
        try {
          console.log(`嘗試從 Topic weather_${oldTownshipCode} 取消訂閱 Token ${oldFcmToken.substring(0, 10)}...`);
          await messaging.unsubscribeFromTopic(oldFcmToken, `weather_${oldTownshipCode}`);
          console.log('成功取消舊的訂閱');
        } catch (e) {
          console.error('取消訂閱時發生錯誤 (可忽略):', e.message);
        }
      }

      // 2. 如果有新的鄉鎮代碼，訂閱新的
      if (townshipCode) {
        try {
          console.log(`嘗試將 Token ${fcmToken.substring(0, 10)}... 訂閱到 Topic weather_${townshipCode}`);
          await messaging.subscribeToTopic(fcmToken, `weather_${townshipCode}`);
          console.log('成功訂閱新的 Topic');
        } catch (e) {
           console.error('訂閱新 Topic 時發生錯誤:', e.message);
           // 如果訂閱失敗，則拋出錯誤，讓整個請求失敗
           throw e;
        }
      }
    } else {
      console.log('資料與現有紀錄相同，無需更新訂閱。');
    }
    
    // 更新 Firestore 中的資料
    console.log('正在更新 Firestore 中的 Token 資料...');
    await userRef.set({
      uid: uid,
      fcmToken: fcmToken,
      townshipCode: townshipCode,
      lastUpdated: admin.firestore.FieldValue.serverTimestamp()
    });
    console.log('Firestore 資料更新成功。');
    
    // 回傳成功訊息
    res.status(200).send({ message: 'FCM registration and topic subscription updated successfully' });

  } catch (error) {
    console.error(`FCM 註冊流程發生嚴重錯誤:`, error);
    res.status(500).send({ error: `FCM registration failed: ${error.message}` });
  }
});


// --- 啟動伺服器 ---
const PORT = 7800;
app.listen(PORT, '0.0.0.0', () => {
  console.log(`Node.js FCM server is running on http://0.0.0.0:${PORT}`);
});
